test/tlsconnection.js

```diff
@@ -0,0 +1,1160 @@
+      await server.recv(clientHello);
+    }, TLSError, /ILLEGAL_PARAMETER/);
+  });
+
+  it('errors if ClientHello has more PSK binders than PSK identities', async () => {
+    const clientHello = await testHelpers.makeClientHelloRecord({
+      extensions: [
+        testHelpers.makeSupportedVersionsExtension([0x0304]),
+        testHelpers.makePskKeyExchangeModesExtension([0x00]),
+        testHelpers.makePreSharedKeyExtension([TEST_VECTORS.PSK_ID], [zeros(32), zeros(32)])
+      ]
+    });
+    await assert.throwsAsync(async () => {
+      await server.recv(clientHello);
+    }, TLSError, /ILLEGAL_PARAMETER/);
+  });
+
+  it('errors if ClientHello has a binder that is too short', async () => {
+    const clientHello = await testHelpers.makeClientHelloRecord({
+      extensions: [
+        testHelpers.makeSupportedVersionsExtension([0x0304]),
+        testHelpers.makePskKeyExchangeModesExtension([0x00]),
+        testHelpers.makePreSharedKeyExtension([TEST_VECTORS.PSK_ID], [zeros(24)])
+      ]
+    });
+    await assert.throwsAsync(async () => {
+      await server.recv(clientHello);
+    }, TLSError, /DECODE_ERROR/);
+  });
+
+  it('errors if ClientHello has data after the extensions', async () => {
+    const clientHello = await testHelpers.makeClientHelloRecord({
+      extensions: [
+        testHelpers.makeSupportedVersionsExtension([0x0304]),
+        testHelpers.makePskKeyExchangeModesExtension([0x00]),
+        testHelpers.makePreSharedKeyExtension([TEST_VECTORS.PSK_ID], [zeros(32)])
+      ],
+      trailer: zeros(12)
+    });
+    await assert.throwsAsync(async () => {
+      await server.recv(clientHello);
+    }, TLSError, /DECODE_ERROR/);
+  });
+
+  it('errors if the ClientHello contains duplicate extensions', async () => {
+    const clientHello = await testHelpers.makeClientHelloRecord({
+      extensions: [
+        testHelpers.makeSupportedVersionsExtension([0x0304]),
+        testHelpers.makePskKeyExchangeModesExtension([0x00]),
+        testHelpers.makeSupportedVersionsExtension([0x0304]),
+        testHelpers.makePreSharedKeyExtension([TEST_VECTORS.PSK_ID], [zeros(32)])
+      ],
+    }, TEST_VECTORS.PSK_ID);
+    await assert.throwsAsync(async () => {
+      await server.recv(clientHello);
+    }, TLSError, /DECODE_ERROR/);
+  });
+});
+
+describe('the ClientConnection class', () => {
+  let client, CLIENT_SENT;
+
+  beforeEach(async () => {
+    CLIENT_SENT = [];
+    const RANDOM_VALUES = [TEST_VECTORS.CLIENT_RANDOM, TEST_VECTORS.SESSION_ID];
+    sinon.stub(crypto, 'getRandomValues').callsFake(arr => {
+      arr.set(RANDOM_VALUES.shift());
+    });
+    client = await ClientConnection.create(TEST_VECTORS.PSK, TEST_VECTORS.PSK_ID, data => CLIENT_SENT.push(data));
+  });
+
+  afterEach(() => {
+    crypto.getRandomValues.restore();
+  });
+
+  it('calls crypto.getRandomValues to generate a strong random salt and session id', async () => {
+    assert.equal(crypto.getRandomValues.callCount, 2);
```

Comment:
Note that getRandomValues is not required for the session ID.  NSS has an awful kludge for that that doesn't involve looking at the random.

I don't mind a little white-box testing.  But you might want to check that both calls ask for 32 octets.